src/process_utility.c

```diff
@@ -121,41 +135,158 @@ process_copy(Node *parsetree, const char *query_string, char *completion_tag)
-		}
-	}
+	if (!OidIsValid(relid))
+		return false;
 
-	if (stmt->is_from && ht != NULL)
-		timescaledb_DoCopy(stmt, query_string, &processed, ht);
-	else
-		DoCopy(stmt, query_string, &processed);
+	hcache = hypertable_cache_pin();
+	ht = hypertable_cache_get_entry(hcache, relid);
+
+	if (ht == NULL)
+	{
+		cache_release(hcache);
+		return false;
+	}
 
+	executor_level_enter();
+	timescaledb_DoCopy(stmt, query_string, &processed, ht);
 	executor_level_exit();
+
 	processed += executor_get_additional_tuples_processed();
 
 	if (completion_tag)
 		snprintf(completion_tag, COMPLETION_TAG_BUFSIZE,
 				 "COPY " UINT64_FORMAT, processed);
 
-	if (NULL != hcache)
-		cache_release(hcache);
+	cache_release(hcache);
+
+	return true;
+}
+
+typedef void (*process_chunk_t) (Oid chunk_relid, void *arg);
+
+/*
+ * Applies a function to each chunk of a hypertable.
+ *
+ * Returns the number of processed chunks, or -1 if the table was not a
+ * hypertable.
+ */
+static int
+foreach_chunk(RangeVar *rv, process_chunk_t process_chunk, void *arg)
+{
+	Oid			relid = hypertable_relid(rv);
+	List	   *chunks;
+	ListCell   *lc;
+	int			n = 0;
+
+	if (!OidIsValid(relid))
+		/* Not a hypertable */
+		return -1;
+
+	chunks = find_inheritance_children(relid, NoLock);
+
+	foreach(lc, chunks)
+	{
+		process_chunk(lfirst_oid(lc), arg);
+		n++;
+	}
+
+	return n;
+}
+
+typedef struct VacuumCtx
+{
+	VacuumStmt *stmt;
+	bool		is_toplevel;
+} VacuumCtx;
+
+/* Vacuums a single chunk */
+static void
+vacuum_chunk(Oid chunk_relid, void *arg)
+{
+	VacuumCtx  *ctx = (VacuumCtx *) arg;
+	char	   *relschema = get_namespace_name(get_rel_namespace(chunk_relid));
+	char	   *relname = get_rel_name(chunk_relid);
+
+	ctx->stmt->relation->relname = relname;
+	ctx->stmt->relation->schemaname = relschema;
+	ExecVacuum(ctx->stmt, ctx->is_toplevel);
+}
+
+/* Vacuums each chunk of a hypertable */
+static bool
+process_vacuum(Node *parsetree, ProcessUtilityContext context)
+{
+	VacuumStmt *stmt = (VacuumStmt *) parsetree;
+	VacuumCtx	ctx = {
+		.stmt = stmt,
+		.is_toplevel = (context == PROCESS_UTILITY_TOPLEVEL),
+	};
+
+	if (stmt->relation == NULL)
+		/* Vacuum is for all tables */
+		return false;
+
+	PreventCommandDuringRecovery((stmt->options & VACOPT_VACUUM) ?
+								 "VACUUM" : "ANALYZE");
+
+	return foreach_chunk(stmt->relation, vacuum_chunk, &ctx) >= 0;
+}
+
+static void
+reindex_chunk(Oid chunk_relid, void *arg)
+{
+	ReindexStmt *stmt = (ReindexStmt *) arg;
+
+	if (stmt->kind == REINDEX_OBJECT_TABLE)
+	{
+		char	   *relschema = get_namespace_name(get_rel_namespace(chunk_relid));
+		char	   *relname = get_rel_name(chunk_relid);
+
+		stmt->relation->relname = relname;
+		stmt->relation->schemaname = relschema;
+		ReindexTable(stmt->relation, stmt->options);
+	}
+	else if (stmt->kind == REINDEX_OBJECT_INDEX)
+	{
+		/*
+		 * Currently not supported. Need to look up the chunk index that
+		 * corresponds to the hypertable's index.
+		 */
+		/* ReindexIndex(stmt->relation, stmt->options); */
+	}
```

Comment:
Yeah, I guess. I figured that if I raise an error here, the command won't execute on the main table either. But, thinking about it, that's not a big deal since that table cannot have any tuples in any case, thus reindexing on that table is pointless.